src/software/new_geom/util/projection.h

```diff
@@ -0,0 +1,175 @@
+                                                        Segment segment2);
+
+/**
+ * Function merges two parameter segments into a single segment ONLY IF one of the
+ * segments is contained ENTIRELY within the other
+ *
+ * @param segment1 : first segment
+ * @param segment2 : second segment
+ * @return Segment : The longer segment ONLY IF one of the segments is contained entirely
+ * within the other
+ * @return nullopt : One of the segments is not fully contained in the other, OR the
+ * segments are not collinear
+ */
+std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment segment2);
+
+/**
+ * Function projects Circle objects onto a Segment with respect to an origin Point
+ *
+ *               projected Segment
+ * *______X---------------------------X___________________*  <-- reference Segment
+ *          .                        .
+ *           .                      .
+ *            .                    .
+ *             .                  .
+ *              .                .
+ *               .              .
+ *                .   Circle   .
+ *                 .  /----\  .
+ *                  . |    | .
+ *                   .\----/.
+ *                    .    .
+ *                     .  .
+ *                      ..
+ *                       X
+ *                 Reference Origin
+ *
+ * @param segment : The reference Segment to project onto
+ * @param circles : The vector of circles to project onto the reference Segment
+ * @param origin : The Point origin to calculate the projection of the Circles from
+ *
+ * @return vector<Segment> : All of the projections of the Circles onto the reference
+ * Segment
+ */
+std::vector<Segment> projectCirclesOntoSegment(Segment segment,
+                                               std::vector<Circle> circles, Point origin);
+
+/**
+ * Function that returns a Segment of the 'Empty' space between the Vector of Segments AND
+ * within the parent_segment.
+ *
+ * NOTE: All segments must be COLLINEAR and must NOT occupy any of the same space
+ *       Additionally, the parent segment completely contain ALL segments in the same
+ * space
+ *
+ *  *-------------------------------------------------* <-- Parent segment
+ *        *xxxxxxxx*          *xxxxxxxxxx*    *xxxx*    <-- Occupied space (input
+ * Segments)
+ *  *_____*         *_________*          *____*    *__* <-- Empty space (output Segments)
+ *
+ * @param segments : The vector of Segments that represent the 'occupied' space. These
+ * will be used to find the Segments representing the 'Empty' space
+ * @param parent_segment : The Segment representing the entire linear space to calculate
+ * the 'Empty' space Segments from
+ *
+ * @return vector<Segment> : All of the 'Empty' space Segments enclosed in the parent
+ * segment
+ */
+std::vector<Segment> getEmptySpaceWithinParentSegment(std::vector<Segment> segments,
+                                                      Segment parent_segment);
```

Comment:
Most of these functions don't seem to have anything to do with projection?

Reply:
yeah I agree, moved them out

Reply:
What do we think about moving these functions into `projection_impl.{h,cpp}` files, assuming that this are just used for projection? We can put `projection_impl.h` in the `srcs` for the `projection` bazel target so that nothing else can see it,`projection_test` will just need to include it in it's `srcs` as well. 

ie.

```
bazel_library(
        name = "projection",
        srcs = [
            "projection_impl.h"
            "projection_impl.cpp"
            "projection.cpp"
        ],
        hdrs = [
            "projection.h"
        ],
        )

bazel_test(
        name = "projection_test",
        srcs = [
            "projection_test.cpp",
            "projection_impl.h"
        ],
        deps = [
            "@gtest//:gtest_main"
            ":projection"
        ],
        )
```

Reply:
We can do that. Just a note that mergeOverlappingSegments is used by calc_best_shot

Reply:
done